### PR TITLE
nixos/modules/networking/wg-quick Add wg-quick options support

### PR DESCRIPTION
**Pull request by exi**
This is an implementation of wireguard support using wg-quick config
generation.

This seems preferable to the existing wireguard support because
it handles many more routing and resolvconf edge cases than the
current wireguard support.

I'm aware that there are situations where you want the explicit control over wireguard provided by the existing networking.wireguard module but other distributions are also shipping wg-quick@.service units to make it very hard to do the wrong thing.

Our hand baked wireguard setups already diverge from wg-quick in significant ways and for the worse.
Wrong ipv6 routing and missing dns handling make it very easy to accidentally pass traffic past wireguard even if allowedIps is set to ::/0.

I'm not advocating to remove networking.wireguard but i would like to offer a way to get the "easy and secure  by default" behavior that wg-quick provides.

It also includes workarounds to make key files work.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by exi** [comment]
@fpletz pretty sure you will have some opinion on this as well.

**Comment by exi** [comment]
@zx2c4 I say you doing some similar work only after i opened this PR.
Any thoughts?

I feel like using wg-quick to get the default and safe behaviour around routing/dns/etc. is not a bad approach.
This should help to avoid "works on my machine" issues between distros.

**Comment by NinjaTrappeur** [comment]
Hey @exi, are you still working on this?

**Comment by exi** [comment]
@NinjaTrappeur sure but nobody responded to my PR

**Comment by zx2c4** [comment]
Sorry for the delay in responding here. I haven't read the commit yet but I vaguely recall being poked, at some point, about my opinion replacing the nixos backend with a wg-quick powered backend.

My general feeling about this is that wg-quick is a pile of hacks, and that distributions that already have control over the networking stack can do much better with something custom and focused. I'm happy to help the nix module if there are particular problems with it that need fixing.

Alternatively, if man power is short and you do prefer just relying on the wg-quick stuff, I guess that's fine too, but it probably won't be as slick as the optimal nix-centric integration.

There was some discussion in https://github.com/NixOS/nixpkgs/pull/31250 IIRC.

**Comment by NinjaTrappeur** [comment]
See also https://github.com/NixOS/nixpkgs/issues/52411

**Comment by exi** [comment]
@globin ping

File: nixos/modules/services/networking/wg-quick.nix [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+      table = mkOption {
+        example = "main";
+        default = null;
+        type = with types; nullOr str;
+        description = ''
+          The kernel routing table to add this interface's
+          associated routes to. Setting this is useful for e.g. policy routing
+          ("ip rule") or virtual routing and forwarding ("ip vrf"). Both numeric
+          table IDs and table names (/etc/rt_tables) can be used. Defaults to
+          "main".
```
Comment (asymmetric) [inline nixos/modules/services/networking/wg-quick.nix]: Should the default be declared in the option's `default` instead?

**Comment by globin** [comment]
I'd prefer this to use the same option names as in `networking.wireguard`.

Otherwise I think this preferable to `networking.wireguard` for simple setups and should be included for the time being. Most probably this will become obsolete when we switch to networkd by default, but this can easily be deprecated in that case.

**Comment by exi** [comment]
@globin I would prefer to keep the names identical to every wg-quick config out there but i don't really have a strong opinion. If you still prefer to make it match our own special snowflake names i could do that as well.
Just tell me which one you prefer.

**Comment by globin** [comment]
Ah then fine by me, when deprecating we can easily add the renames.

**Comment by asymmetric** [comment]
Would it make sense to backport this to 19.03?

**Comment by andrew-d** [comment]
I would love to see this backported; it's a huge quality-of-life improvement for using Wireguard, and it doesn't break anything.

File: nixos/modules/services/networking/wg-quick.nix [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+
+
+  ###### implementation
+
+  config = mkIf (cfg.interfaces != {}) {
+    boot.extraModulePackages = [ kernel.wireguard ];
+    environment.systemPackages = [ pkgs.wireguard-tools ];
+    # This is forced to false for now because the default "--validmark" rpfilter we apply on reverse path filtering
+    # breaks the wg-quick routing because wireguard packets leave with a fwmark from wireguard.
+    networking.firewall.checkReversePath = false;
```
Comment (c0bw3b) [inline nixos/modules/services/networking/wg-quick.nix]: I don't think a service module should **silently** change the firewall behaviour

**Comment by asymmetric** [comment]
Should this module configure the firewall to allow incoming UDP packets on the specified port?

**Comment by c0bw3b** [comment]
Not automatically, the policy is that opening ports _should_ always be an explicit user configuration (SSH being the only exception to avoid locking oneself out of a remote machine).

The module description _could_ indicate that additional firewall config is needed.
BTW I have not tested wg but I'm not sure I see why wg marks are not considered valid marks. Does setting `checkReversePath` to loose change anything?

**Comment by exi** [comment]
> The module description _could_ indicate that additional firewall config is needed.
> BTW I have not tested wg but I'm not sure I see why wg marks are not considered valid marks. Does setting `checkReversePath` to loose change anything?

The reason it is not considered a valid mark is because we receive a packet from the network without a mark but sent the response with a mark because wg-quick uses firewall marks only for outgoing packages because it makes it really easy to bypass the default routing table.
It needs to bypass the default routing table because you can configure wireguard to be your default route. In this case, all outgoing packets would use the wireguard default route except the ones with the firewall mark.

However this causes us to use a different return path from the perspective of rp_filter, which considers (incoming, nomark) and (outgoing, mark) as two different paths as long as --validmark is set. And if it considers these to be two different paths, it will drop the packets.

File: nixos/modules/services/networking/wg-quick.nix (hunk 2 above)
Comment (c0bw3b) [reply nixos/modules/services/networking/wg-quick.nix]: I'd like to request @fpletz opinion on this :
is it something we should keep as-is for the next release?

**Comment by exi** [comment]
> 
> 
> Should this module configure the firewall to allow incoming UDP packets on the specified port?

Not sure, for wg-quick we have outgoing packets first. So incoming packets would fall into a RELATED state and i presume (without checking) that related packets are already allowed.

**Comment by asymmetric** [comment]
Is it not supposed to be used on the server?

**Comment by exi** [comment]
> Is it not supposed to be used on the server?

Yes you are right. We may need some complex logic to figure out what port to open where.
I honestly would leave that to the user to explicitly open a port. I don't think this is something we could get right by default.

**Comment by asymmetric** [comment]
Yeah I don't think we should open a port either, as it seems to be against the general guidelines.

Maybe just mention in the module that whatever port one specifies should be opened for a server-like peer.

**Comment by andrew-d** [comment]
Maybe have a `configureNetworking` option that defaults to `false`?  That way, it's untouched by default, but if you're just doing the simple case of enabling a Wireguard server, you can set one option and it opens the right `listenPort` (if given), sets `checkReversePath = false`, etc. etc.? If we wanted to get really fancy, we could also create a systemd unit that waits for the Wireguard interface to come up, dynamically looks up the listening port (for cases where you don't specify one), and then inserts a firewall rule?

I dunno; no strong opinions here, but I think it's worth optimizing for the easy case here; most people that set up a Wireguard server will need to set the rpfilter/port networking options anyway.

Other thoughts:
- `mkDefault` for `checkReversePath` in this module?
- `assertions` to ensure that it's set in the cases where we know that it's needed?

**Comment by c0bw3b** [comment]
> Maybe just mention in the module that whatever port one specifies should be opened for a server-like peer.

Yes that's the usual approach for modules that needs some firewall rules : description should explain it has to be configured by the user. Then it's left to the user to open the right port on the right interfaces (or all interfaces) according to his own setup.

> * `assertions` to ensure that it's set in the cases where we know that it's needed?

That could be an extra precaution.

**Pull request merged**